### DISCUSSION TITLE
Integrate new feature "My Trips"

### DISCUSSION THREAD

**Discussion by bredmich** [bimmerconnected/bimmer_connected#500] [Feature request]
Hi there,

the MyBMW app received a new update and can now show recent trips of the car.
Could you have a look if this data can be extracted?

The new feature needs to be activated in the data privacy settings before any data will be stored.

![grafik](https://user-images.githubusercontent.com/15615102/205124188-0a951b04-2080-424b-8c03-5caf6964d904.png)

kind regards
Michael

**Comment by HuChundong** [comment]
it's easy.

**Comment by sincze** [comment]
@gerard33 , @rikroe , after using this feature in the myBMW app, would this be possible to integrate in bimmerconnected? That way I can parse the results to my administration more easily ;-0

**Reply by gerard33** [reply]
Check #569

**Reply by bredmich** [reply]
That‘s not the same, as far as i can tell.
The last trips feature is not limited to Hybrid and BEV vehicles.

**Comment by sincze** [comment]
> Check #569 

This seem to be referencing to charging sessions (energy), that works excellent indeed.
![image](https://github.com/bimmerconnected/bimmer_connected/assets/5776333/938cec6a-6a19-44c1-9481-f6098e84a2fd)

I was talking about TRIPS. Example:

Home -> Work
Vertrek: Plataanlaan 2, Utrecht -> Destination: CentraalStation, Amsterdam 66km

**Comment by davidovitch** [comment]
From PR #348 I can see that one could access the My Trips data at some point, but that an updated API removed support for that. The My BMW app is clearly able to extract this data from the My BMW account. Has anyone been looking into that recently by any chance? If not, would the next step be to reverse engineer the API the app is using as described at https://bimmer-connected.readthedocs.io/en/latest/development/reverse_engineering_mybmw.html?

Note that I created a ticket with BMW support requesting that, as a bare minimum, they include the My Trip data in the BMW CarData archive because that is currently not the case. I'll report any updates on that here.

**Reply by davidovitch** [reply]
BMW Support confirmed that it is not possible to export the "My Trips" data. They forwarded this feature request to the developers but did not indicate if they were actually considering this or not.

**Reply by rikroe** [reply]
Both my car and the demo vehicles don't feature the last trips functionality. Therefore I cannot help right now and need somebody to provide the required URL, header and responses.

**Reply by davidovitch** [reply]
> Both my car and the demo vehicles don't feature the last trips functionality. Therefore I cannot help right now and need somebody to provide the required URL, header and responses.

How can you figure what URL is being queried? I have seen https://bimmer-connected.readthedocs.io/en/latest/development/reverse_engineering_mybmw.html but setting that environment up seems a pretty lengthy process (I don't have a rooted Android phone). Do you know of any other way?

**Reply by davidovitch** [reply]
I have not tried it yet, but it seems that https://github.com/TA2k/ioBroker.bmw a few months ago added something to get the trips (see [here](https://github.com/TA2k/ioBroker.bmw/blob/a490aabba5709c560626c225ee3a66d4bb97049f/main.js#L543)). It seems to URL looks something like `https://cocoapi.bmwgroup.com/eadrax-suscs/v1/vehicles/sustainability/widget`. I'll try to figure out how the response looks like and share that in a new issue and/or a PR if I manage to implement a prototype.

**Reply by davidovitch** [reply]
I got now something that works based on the `updateTrips` function from [ioBroker.bmw](https://github.com/TA2k/ioBroker.bmw/blob/a490aabba5709c560626c225ee3a66d4bb97049f/main.js#L543). It only returns the last trip but it is a good start I would say. It looks something like this:

```python
from bimmer_connected.api.client import MyBMWClient
from bimmer_connected.account import MyBMWAccount
from bimmer_connected.api.regions import Regions

account = MyBMWAccount(user, pwd, Regions.REST_OF_WORLD)
await account.get_vehicles()
vehicle = account.get_vehicle(vin)

TRIP_URL = 'https://cocoapi.bmwgroup.com/eadrax-suscs/v1/vehicles/sustainability/widget'
async def get_last_trip(account, vehicle):
    fetched_at = datetime.datetime.now(datetime.timezone.utc)
    async with MyBMWClient(account.config) as client:
        error_count = 0
        headers = {**client.generate_default_header(vehicle.brand),
                   "bmw-vin": vehicle.vin,
                   "x-gcid": account.gcid,
                   "authorization": f"Bearer {account.config.authentication.access_token}",
                   "accept-language": 'de-DE',
                   "host": 'cocoapi.bmwgroup.com',
                  }
        print('header:', headers)
        state_response = await client.get(
            TRIP_URL,
            headers=headers,
        )
        response = state_response.json()
    return response
```

Note that it also requires to save `gcid` since that for some reason this is not done for rest-of-world and US accounts in `accounts.py`.

The response looks like this:
```python
{'status': 'Success', 'vin': 'XYZ', 'gcid': 'XYZ', 'lastTrip': {'id': 'ab1cf6f0-5840-4c00-ba8c-99364c7aa221', 'distance': {'distance': 45.2, 'distanceUnit': 'km'}, 'duration': {'duration': 48, 'durationUnit': 'M'}, 'electricConsumption': {'averageConsumption': 17.2, 'averageUnit': 'kWh100Km'}}, 'monthly': {'totalTrips': 15, 'totalDistance': 396.5, 'totalDistanceUnit': 'km', 'totalElectricConsumption': {'averageConsumption': 18.7, 'averageUnit': 'kWh100Km'}}, 'lastUpdatedAt': '2024-02-06T19:40:11.2334331Z'}
```

@rikroe I have looked at creating a PR from my fork but I realised this package has quite some stuff to structure the data response in a specific manner that I am not sure I really get. Do you have any specific idea on how you would like to integrate a "Trips" data structure into this package?

**Reply by rikroe** [reply]
Hi, cool! If there is an example available, that makes it much easier.
Yes, I know that the library itself is quite complex as we're parsing and converting data from the API.

For the Trips API I would suggest to enhance the **remote service** functions and implement it as a remote service that for a first draft only returns the API result as JSON.

I would not query this API every 5 minutes with the other status information, as we do have quite a lot of issues with Too Many Request errors by the upstream APIs.
So as a remote function, a user could use a Home Assistant Service or CLI call at the times he needs.

Maybe I can have a look at the weekend, I'm quite busy at the moment.

**Reply by davidovitch** [reply]
That's helpful advice, thanks. I will start with creating a PR that implements this in `remote_services.py` and we can take it from there. I would really like to pull out a year worth of trips from my account so I am pretty motivated to get this to work :-) I was hoping BMW would add the option to either save it from the app to some sort of file, or at least include in the BMW Data Archive. They really should do the latter as a minimum since this is pretty sensitive and private GDPR covered data.

**Reply by empe12** [reply]
Hi! Did anyone get any further on this?
I'm currently driving a Model Y and use teslamate - my next car might be an i5 (beginning of 2025). In terms of statistics etc. teslamate is amazing, I'd love to have something like it for the BMW. Especially the trips part helps a lot for later reference.
